core/src/main/java/org/infinispan/container/impl/EntryFactoryImpl.java

```diff
@@ -62,34 +66,110 @@ public void init() {
    }
 
    @Override
-   public final void wrapEntryForReading(InvocationContext ctx, Object key, int segment, boolean isOwner) {
+   public final CompletionStage<Void> wrapEntryForReading(InvocationContext ctx, Object key, int segment, boolean isOwner) {
       if (!isOwner && !isL1Enabled) {
-         return;
+         return CompletableFutures.completedNull();
       }
+      CompletionStage<Void> returnedStage = CompletableFutures.completedNull();
       CacheEntry cacheEntry = getFromContext(ctx, key);
       if (cacheEntry == null) {
-         cacheEntry = getFromContainerForRead(key, segment, isOwner);
+         InternalCacheEntry readEntry = getFromContainer(key, segment);
+         CompletionStage<Boolean> expiredStage = null;
+         if (readEntry == null) {
+            if (isOwner) {
+               cacheEntry = NullCacheEntry.getInstance();
+            }
+         } else if (isOwner || readEntry.isL1Entry()) {
+            cacheEntry = readEntry;
+            expiredStage = handlePossibleExpiredEntry(readEntry, false);
+         }
 
          if (cacheEntry != null) {
-            // With repeatable read, we need to create a RepeatableReadEntry as internal cache entries are mutable
-            // Otherwise we can store the InternalCacheEntry directly in the context
-            if (useRepeatableRead) {
-               MVCCEntry mvccEntry = createWrappedEntry(key, cacheEntry);
-               mvccEntry.setRead();
-               cacheEntry = mvccEntry;
+            if (expiredStage != null) {
+               if (!CompletionStages.isCompletedSuccessfully(expiredStage)) {
+                  CacheEntry finalEntry = cacheEntry;
+                  returnedStage = expiredStage.thenApply(expired -> {
+                     if (expired == Boolean.FALSE) {
```

Comment:
nitpick: you could avoid creating the lambda in `handlePossibleExpiredEntry()` if the `touch()` method is invoked here. 
(or vice-versa, i.e., the `handlePossibleExpiredEntry()` adds the entry to the context)

Reply:
I would also have to change the write portion as well to handle this. I figure that this lambda is only going to be used so infrequently since it only happens if the stage is not complete (which requires the entry to actually be expired).

Reply:
ok :+1: